# Practice Case MongoDB with pymongo
Pada practice case ini akan dibuat suatu collection (tabel) pada cluster yang telah ditentukan. Cluster yang digunakan adalah:

- mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true& w=majority (**cluster 1**)
- mongodb+srv://admin1234:12345@cluster0-miqju.gcp.mongodb.net/test?retryWrites=true&w=majority (**cluster 2**)

Cluster 1 digunakan untuk melihat collection movies. Sementara cluster 2 digunakan untuk  melihat collection movies_initial dan menyimpan collection yang akan dibuat.

Ketentuan dalam membuat collection yang diinginkan adalah sebagai berikut:
1. Membuat collection dengan nama clean_movies_tara.
2. Collection tersebut disimpan pada database sample_mflix pada cluster 2.
3. Values dari collection yang akan dibuat harus sama persis dengan collection movies_initial.
4. Collection movies_initial yang digunakan ada pada database sample_mflix pada cluster 2. 

Validasi untuk collection clean_movies_tara dengan parameter sebagai berikut:
1. Semua document pada clean_movies dan movie sama.
2. Banyak document pada clean_movies dan movie sama.
3. Semua fields pada clean_movies ada pada movie.
4. Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama.

## 1. Install librarys
Librarys pymongo dan pymongo[srv] perlu diinstall pada google colab.

In [0]:
# !pip install pymongo
# !pip install pymongo[srv]

## 2. Load Librarys

In [0]:
# import pandas for show the document in the form of DataFrame
import pandas as pd

# import MongoClient to connect with the cluster
import pymongo
from pymongo import MongoClient

# import pprint to make the output more pretty and sistematics
import pprint

# import clear_output to clean the output from the running code before than repalce with the last running code
from IPython.display import clear_output

## 3. Connect to the Clusters

- **Cluster 1 : userstudent**

In [3]:
# Connect to cluster 1 : userstudent
client_userstudent = MongoClient('mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majorit')
print(client_userstudent)

MongoClient(host=['cluster0-shard-00-01-nnbxe.gcp.mongodb.net:27017', 'cluster0-shard-00-02-nnbxe.gcp.mongodb.net:27017', 'cluster0-shard-00-00-nnbxe.gcp.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majorit', authsource='admin', replicaset='Cluster0-shard-0', ssl=True)


- **Cluster 2 : admin1234**

In [4]:
# Connect to cluster 2 : admin1234
client_admin1234 = MongoClient('mongodb+srv://admin1234:12345@cluster0-miqju.gcp.mongodb.net/test?retryWrites=true&w=majority')
print(client_admin1234)

MongoClient(host=['cluster0-shard-00-00-miqju.gcp.mongodb.net:27017', 'cluster0-shard-00-01-miqju.gcp.mongodb.net:27017', 'cluster0-shard-00-02-miqju.gcp.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='Cluster0-shard-0', ssl=True)


Cluster yang digunakan adalah cluster userstudent dan admin1234. Cluster userstudent tidak dapat menampung 24 collection baru, maka dibuatlah cluster admin1234 untuk memuat 24 collection baru. Ke-24 collection tersebut merupakan hasil collection baru yang telah dibuat oleh 24 orang peserta ADSB Batch 2.

## 4. Data Understanding

## 4.1. Show all databases in the used clusters

- **Database in cluster userstudent**

In [5]:
# Show all database in cluster userstudent
print(client_userstudent.list_database_names())

['sample_mflix', 'admin', 'local']


- **Database in cluster admin1234**

In [6]:
# Show all database in cluster client2
print(client_admin1234.list_database_names())

['sample_mflix', 'admin', 'local']


Pada cluster userstudent dan admin1234 hanya terdapat satu database, yaitu sample_mflix. Database admin dan local adalah default template database pada setiap cluster pada mongoDB.

## 4.2. Define the used database and collections

In [0]:
# define database dan collection to use
db_mflix1 = client_userstudent['sample_mflix']
db_mflix2 = client_admin1234['sample_mflix']

coll_movies = db_mflix1.movies
coll_moviesinitial = db_mflix2.movies_initial

- **Collection** yang digunakan untuk membuat clean_movies_tara adalah **movies** dan **movies_initial**. Movies digunakan sebagai acuan dalam membuat clean_movies_tara. Dimana clean_movies_tara harus sama persis dengan movies. Sementara value untuk clean_movies_tara diambil dari movies_initial.
- Namun kedua collection yang digunakan tersebut terletak pada dua cluster yang berbeda. Movies terletak pada database sample_mflix di cluster userstudent. Sementara movies_initial terletak pada database sample_mflix di cluster admin1234.
- Maka statement tersebut mampu menjawab mengapa digunakan dua cluster untuk membuat clean_movies_tara.

## 4.3. Show all collections in used database

In [8]:
# Show all collection in used database from cluster client
pprint.pprint(db_mflix1.list_collection_names())

['clean_movies_dwilaras',
 'clean_movies_nurlailiis',
 'clean_movies_devita',
 'movies_initial',
 'clean_movies_frans',
 'clean_movies_tara',
 'clean_movies_tunggal',
 'clean_movies',
 'users',
 'clean_movies_nurrimah',
 'movies_testtopik',
 'movies_scratch',
 'clean_movies_topik',
 'movies',
 'clean_movies_afifah',
 'clean_movies_andreas',
 'clean_movies_indra_1']


In [9]:
# Show all collection in used database from cluster client
pprint.pprint(db_mflix2.list_collection_names())

['clean_movies_Putrisqiana',
 'clean_movies_tara',
 'clean_movies_bunga',
 'clean_movies_zumar',
 'clean_movies_jul',
 'movies_initial',
 'clean_movies_putrisqiana',
 'clean_movies_firdaus',
 'clean_movies_dwilaras',
 'clean_movies_nurlailiis',
 'clean_movies_faizah']


- Dari code di atas terlihat bahwa terdapat dua cluster untuk menyimpan 24 collection baru dengan nama clean_movies_(nama). Pada cluster 1 terdapat collection movies dan movies_initial. Sementara pada cluster 2 terdapat collection movies_initial. Untuk memudahkan dalam pembuatan collection baru, maka digunakan movies collection dari cluster 1 dan movies_initial dari cluster 2.

## 4.4. Show one document in the movies collection

- **Show document from movies collection in list**

In [10]:
# Show document from movies collection in list
pprint.pprint(list(coll_movies.find({}).limit(1)))

[{'_id': ObjectId('573a1390f29313caabcd5a93'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
  'countries': ['USA'],
  'directors': ['Reginald Barker',
                'Thomas H. Ince',
                'Raymond B. West',
                'Walter Edwards',
                'David Hartford',
                'Jay Hunt',
                'J. Parker Read Jr.'],
  'fullplot': 'Allegorical film about peace. A king starts a war, many of the '
              'women are against it, people are pressed into service. A count '
              'has constructed a submarine and gets the order to sink an ocean '
              'liner, that is also carrying - supposedly - ammunition for the '
              'enemy. The count refuses to fire the torpedos, and sinks the '
              'submarine. He survives, but in a limbo between death and life '
              'where he meets Jesus, who takes him over to preach peace. '
  

- **Show document from movies collection in Dataframe**

In [11]:
# Show document from movies collection in Dataframe
df_movies = pd.DataFrame(list(coll_movies.find({}).limit(2)))
df_movies

_id  ...  year
0  573a1390f29313caabcd5a93  ...  1916
1  573a1390f29313caabcd60e4  ...  1917

[2 rows x 21 columns]

## 4.5. Show one document in movies_initial collection

- **Show document from movies_initial collection in list**

In [74]:
# Show document from movies_initial collection in list
pprint.pprint(list(coll_moviesinitial.find({}).limit(1)))

[{'_id': ObjectId('5daffb1b986c4adedb754af9'),
  'awards': '',
  'cast': 'Carmencita',
  'country': 'USA',
  'director': 'William K.L. Dickson',
  'fullplot': 'Performing on what looks like a small wooden stage, wearing a '
              'dress with a hoop skirt and white high-heeled pumps, Carmencita '
              'does a dance with kicks and twirls, a smile always on her face.',
  'genre': 'Documentary, Short',
  'imdbID': 1,
  'imdbRating': 5.9,
  'imdbVotes': 1032,
  'language': '',
  'lastupdated': '2015-08-26 00:03:45.040000000',
  'metacritic': '',
  'plot': 'Performing on what looks like a small wooden stage, wearing a dress '
          'with a hoop skirt and white high-heeled pumps, Carmencita does a '
          'dance with kicks and twirls, a smile always on her face.',
  'poster': 'http://ia.media-imdb.com/images/M/MV5BMjAzNDEwMzk3OV5BMl5BanBnXkFtZTcwOTk4OTM5Ng@@._V1_SX300.jpg',
  'rating': 'NOT RATED',
  'released': '',
  'runtime': '1 min',
  'title': 'Carmencita',
  'ty

- **Show document from movies_initial collection in Dataframe**

In [13]:
# Show document from movies_initial collection in Dataframe
df_moviesinitial = pd.DataFrame(list(coll_moviesinitial.find({}).limit(2)))
df_moviesinitial

_id  awards  ... writer  year
0  5daffb1b986c4adedb754af9          ...         1894
1  5daffb1b986c4adedb754afa  1 win.  ...         1893

[2 rows x 22 columns]

## 5. Create new collection named clean_movies_tara

## 5.1. Define the imdb_list for match the data in pipeline

In [32]:
# Define the query and projection
query = {}
projection = {'_id' : 0, 'imdb' : 1}

# The number od document in movies collection from imdb field
imdb_movies = list(coll_movies.find(query, projection))

print('The number of document in movies collection : ', len(imdb_movies))

The number of document in movies collection :  23541


In [15]:
# Append all the imdb field from movies collection
imdb_list = []
for i in imdb_movies:
  imdb_list.append(list(i.values())[0])

# Show the top 5 imdb field from movies collection
imdb_list[:5]

[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

In [16]:
# Only take the imdb.id field and make all the imdb.id  as array imdb_id
imdb_id = []
for i in imdb_list:
  imdb_id.append(i['id'])

# Show the fist imdb.id from the array imdb_id
imdb_id[:5]

[6517, 8133, 12512, 13025, 15498]

## 5.2. Create the pipeline for clean_movies_tara & store that collection in cluster admin1234

In [26]:
# drop the collection before and replace with the last running collection
client_admin1234.sample_mflix.drop_collection('clean_movies_tara')

# make the pipeline for clean_movies_tara
pipeline = [
    # construct the field
    {
        '$project' : {
            
            # 1. awards
            'awards': {'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': ['$awards', ' win']}, -2]}, ' ']}, -1]},
                       'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]}, ' ']}, -1]},
                       'text': '$awards'},

            # 2. cast
            'cast' : {'$cond' : [{'$eq' : ['$cast', '']}, '$nonExistinField', '$cast']},
            
            # 3. countries
            'countries' : {'$cond' : [{'$eq' : ['$country', '']}, '$nonExistinField', {'$split' : ['$country', ', ']}]},
            
            # 4. directors
            'directors' : {'$cond' : [{'$eq' : ['$director', '']}, '$nonExistinField', {'$split' : ['$director', ', ']}]},
            
            # 5. fullplot
            'fullplot' : {'$cond' : [{'$eq' : ['$fullplot', '']}, '$nonExistinField', '$fullplot']},
            
            # 6. genres
            'genres' : {'$cond' : [{'$eq' : ['$genre', '']}, '$nonExistinField', {'$split' : ['$genre', ', ']}]},
            
            # 7. imdb
            'imdb' : {'id' : '$imdbID',
                      'rating' : '$imdbRating',
                      'votes' : '$imdbVotes'},

            # 8. languages
            'languages' : {'$cond' : [{'$eq' : ['$language', '']}, '$nonExistinField', {'$split' : ['$language', ', ']}]},
            
            # 9. lastupdated
            'lastupdated' : {'$cond' : [{'$eq' : ['$lastupdated', '']}, '$nonExistinField', '$lastupdated']},
            
            # 10. metacritic
            'metacritic' : {'$cond' : [{'$eq' : ['$metacritic', '']}, '$nonExistinField', '$metacritic']},
            
            # 11. num_mflix_comments
            'num_mflix_comments' : {'$cond' : [{'$eq' : ['$num_mflix_comments', '']}, '$nonExistinField', '$num_mflix_comments']},
            
            # 12. plot
            'plot' : {'$cond' : [{'$eq' : ['$plot', '']}, '$nonExistinField', '$plot']},
            
            # 13. poster
            'poster' : {'$cond' : [{'$eq' : ['$poster', '']}, '$nonExistinField', '$poster']},
            
            # 14. rated
            'rated' : {'$cond' : [{'$eq' : ['$rating', '']}, '$nonExistinField', '$rating']},
            
            # 15. released
            'released': {'$concat' : ['datetime.datetime(',
                                      {'$arrayElemAt':[{'$split': ['$released', '-']}, 0]},
                                      ', ',
                                      {'$arrayElemAt':[{'$split': ['$released', '-']}, 1]},
                                      ', ',
                                      {'$arrayElemAt':[{'$split': ['$released', '-']}, 2]},
                                      ', 0, 0)']  
                        },
            
            # 16. runtime
            'runtime': {'$cond': [{'$eq': ['$runtime', '']}, '$nonExistinField', {'$arrayElemAt':[{'$split': [ '$runtime', ' min' ]}, 0]}]},

            # 17. title
            'title' : {'$cond' : [{'$eq' : ['$title', '']}, '$nonExistinField', '$title']},
            
            # 18. tomatoes
            'tomatoes': {'dvd' : '',
                         'fresh' : '',
                         'lastUpdated' : {'$concat' : ['datetime.datetime(',
                                                       {'$arrayElemAt':[{'$split': ['$lastupdated', '-']}, 0]},
                                                       ', ',
                                                       {'$arrayElemAt':[{'$split': ['$lastupdated', '-']}, 1]},
                                                       ', ',
                                                       {'$arrayElemAt':[{'$split': [{'$arrayElemAt':[{'$split': ['$lastupdated', ' ']}, 0]}, '-']}, 2]},
                                                       ', ',
                                                       {'$arrayElemAt':[{'$split': [{'$arrayElemAt':[{'$split': ['$lastupdated', ' ']}, 1]}, ':']}, 0]},
                                                       ', ',
                                                       {'$arrayElemAt':[{'$split': [{'$arrayElemAt':[{'$split': ['$lastupdated', ' ']}, 1]}, ':']}, 1]},
                                                       ', ',
                                                       {'$arrayElemAt':[{'$split': [{'$arrayElemAt': [{'$split': [{'$arrayElemAt':[{'$split': ['$lastupdated', ' ']}, 1]}, ':']}, 2]}, '.']}, 0]},
                                                       ')']
                                         },
                          'production' : '',
                          'rotten' : '',
                          'viewer' : {'meter' : '',
                                      'numReviews' : '',
                                      'rating' : ''},
                        },
            
            # 19. type
            'type' : {'$cond' : [{'$eq' : ['$type', '']}, '$nonExistinField', '$type']},
            
            # 20. writers
            'writers' : {'$cond' : [{'$eq' : ['$writer', '']}, '$nonExistinField', {'$split' : ['$writer', ', ']}]},
            
            # 21. year
            'year' : {'$cond' : [{'$eq' : ['$year', '']}, '$nonExistinField', '$year']}
             }
    },
    
    # sort the imdb become :  'id', 'rating', 'votes'
    {
        '$match' : {'imdb' : {'$in' : [dict(sorted(imdb_list[x].items(), key = lambda x : x[0])) for x in range(len(imdb_list))]}}
            
    },
    
    # rename the new collection with clen_movies_tara
    {
        '$out' : 'clean_movies_tara'
    }
]

clear_output()
pprint.pprint(list(client_admin1234.sample_mflix.movies_initial.aggregate(pipeline)))

[]


Penjelasan dari ketentuan masing-masing fields pada clean_movies_tara agar sama persis dengan movies collection.
- **awards** : berbentuk list dan terdiri atas 3 elemen, yaitu nominations, text, dan win. Untuk mendapatkan nominations dan wins kita perlu split awards pada movies_initial kemudian mengambil array index 0. Array tersebut akan dimasukkan kedalam win jika array index 1 win, dan sebaliknya. Kemudian untuk elemen text langsung diambil dari awards pada movies_initial.
- **imdb** : berbentuk list dari terdiri atas 3 element, yaitu id, rating, dan votes. Dimana ketiganya diambil dari imdbID, imdbRating, dan imdbVotes pada movies_initial secara berurutan.
- **released** : berbentuk datetime dengan element tahun, bulan, tanggal, jam, dan menit. Nilai dari tahun, bulan, dan tanggal di dapat dari released pada movies_initial, kemudian untuk jam dan menit cukup dituliskan '0, 0' saja. Untuk mendapatkan format yang sesuai, maka dilakukan split pada movies_initial dengan acuan '-'.
- **tomatoes** : berbentuk list terdiri atas 6 element, yaitu dvd, fresh, lastUpdated, production, rotten, dan viewer. Dari ke elemen tersebut hanya lastUpdated saja yang ada nilainya pada movies_initial. Sehingga untuk elemen yang lainnya cukup dikosongkan saja. LastUpdated didapat dari lastupdated pada movies_initial kemudian dijadikan dalam bentuk datetime. Kompoenen datetime yang digunakan terdiri atas tahun, bulan, hari, jam, detik, dan menit. Semua elemen tersebut ada pada lasupdated. Sehingga perlu dilakukan split array kemudian dijadikan dalam bentuk seperti released yang telah dijelaskan sebelumnya. Kemudian untuk viewer berbentuk list yang terdiri atas 3 element, yaitu meter, numReviews, dan rating.
- **writers**, **runtime**, **languages**, **genres**, **directors**, dan **countries**: berbentuk array dengan elemen menyesuaikan banyaknya values pada masing-masing field tersebut di movies_initial. Sebelumnya fields tersebut berbentuk string pada movies_initial. Sehingga perlu dilakukan split dengan acuan ','. Kemudian masing-masing elemen tersebut dimasukkan ke dalam array.
- **cast**, **fullplot**, **lastupdated**, **metacritics**, **num_mflix_comments**, **plot**, **poster**, **rated**, **title**, **type**, **year** : berbentuk string dan integer. Nilainya tunggal sehingga langsung didapatkan dari masing-masing field tersebut pada movies_initial.

<br>
Untuk field pada point 5 dan 6 di atas, perlu diberikan kondisi jika nilainya null / no field pada suatu document maka field-nya akan dihapuskan hanya pada document tersebut. Untuk field lainnya tidak perlu dilakukan karena sudah pasti nilainya akan ada (tidak kosong).

## 5.3. Define clean_movies_tara collection

In [0]:
# define collection clean_movies_tara in client2
coll_tara = db_mflix2.clean_movies_tara

## 5.4. Show one document in clean_movies_tara collection

In [28]:
# Show all document in coll = clean_movies_tara with list
pprint.pprint(list(coll_tara.find({}).limit(1)))

[{'_id': ObjectId('5daffb1b986c4adedb754afa'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': 'Charles Kayser, John Ott',
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': 'datetime.datetime(1893,

In [29]:
# Show all document in coll = clean_movies_tara with DataFrame
df_tara = pd.DataFrame(list(coll_tara.find({}).limit(2)))
df_tara

_id  ...  year
0  5daffb1b986c4adedb754afa  ...  1893
1  5daffb1b986c4adedb754b15  ...  1903

[2 rows x 19 columns]

## 6. Validation

## 6.2. The number of documents in movies and clean_movies_tara are the same

In [68]:
# banyaknya document pada collection movies
imdb_movies = list(coll_movies.find({}))
print('The number of documents in movies collection : ', len(imdb_movies))

The number of documents in movies collection :  23541


In [69]:
# banyaknya document pada collection clean_movies_tara
imdb_tara = list(coll_tara.find({}))
print('The number of documents in clean_movies_tara collection : ', len(imdb_tara))

The number of documents in clean_movies_tara collection :  23539


- Dari code di atas terlihat bahwa **banyaknya document** pada **movies** collection senilai **23541**. Sementara banyaknya document pada **clean_movies_tara** collection senilai **23539**. Sehingga terdapat **selisih 2 document**. Dimana pada clean_movies_tara kekurangan 2 document dari movies collection.
- Selanjutnya akan dilihat apakah sebabnya pada clean_movies_tara collection kekurangan dua document dari movies collection.

In [54]:
# Show the difference documents that didn't include in clean_movies_tara collection
group_tara = list(client_admin1234.sample_mflix.clean_movies_tara.aggregate([{"$sortByCount":"$imdb"}]))
group_movies = list(client_userstudent.sample_mflix.movies.aggregate([{'$project':
                                                                      {'imdb': {'id': '$imdb.id', 'rating': '$imdb.rating', 'votes': '$imdb.votes'}}},
                                                                               {"$sortByCount":"$imdb"}]))


for i in group_movies:
  if i not in group_tara:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 3}


- Dari code tersebut di atas diketahui bahwa **selisih 2 document** ada pada document dengan kriteria value dari **imdbID : 5**, **imdbRating : 6.2**, dan **imdbVotes : 1189**. Pada output di atas terlihat bahwa document dengan keterangan tersebut sejumlah 3 baris. Dengan satu diantaranya sudah masuk kedalam clean_movies_data namun dua lainnya belum masuk.
- Selanjutnya akan dilihat bagaimanakah detail data yang bermasalah sebagai berikut:

In [65]:
# Show all document in movies collection where imdbID = 5
# There are 3 documents in movies collection where imdbID = 5
pprint.pprint(list(coll_movies.find({'imdb.id': 5, 'imdb.rating': 6.2, 'imdb.votes': 1189})))

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released'

- Dari output di atas terlihat bahwa terdapat tiga document yang hampir sama persis. Yang membedakan diantara tiga baris data tersebut terletak pada **genre** nya saja. Salah satunya ber-genre **Short** dan dua lainnya ber-genre **Short dan Long**. Serta dapat dilihat bahwa value pada field lainnya sama persis. Sehingga indikasi yang dapat dikatakan adalah terdapat kekurangan input dua document tersebut pada movies_initial. Sehingga ketika dilakukkan pembuatan pipeline pasti akan terdapat **kekurangan 2 document** apda **clean_movies_tara**.
- Untuk memperkuat argumen tersebut maka akan dilakukan pengecekan pada movies_initial collection. Akan diperlihatkan berapa banyak baris data dengan imdbID = 5.

In [62]:
# Show document in movies_initial collection where imdbID = 5
pprint.pprint(list(coll_moviesinitial.find({'imdbID':5})))

[{'_id': ObjectId('5daffb1b986c4adedb754afa'),
  'awards': '1 win.',
  'cast': 'Charles Kayser, John Ott',
  'country': 'USA',
  'director': 'William K.L. Dickson',
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genre': 'Short',
  'imdbID': 5,
  'imdbRating': 6.2,
  'imdbVotes': 1189,
  'language': '',
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'metacritic': '',
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'poster': '',
  'rating': 'UNRATED',
  'released': '1893-05-09',
  '

- Dari output di atas di dapat bahwa hanya terdapat **satu document** pada **movies_initial** dengan nilai **imdbID : 5**. Sehingga kesimpulan yang dapat diambil adalah terdapat kekurangan document yang belum dimasukkan kedalam movies_initial. Sehingga banyaknya data pada clean_movies_tara tidak akan sama persis dengan movies jika diharapkan data tersebut memiliki value yang sama persis antara keduanya.

## 6.3. All the fields in movies and clean_movies_tara are the same

In [49]:
# Show all fields in movies collection
field_movies = [list(doc.keys()) for doc in coll_movies.find({})]
field_movies = set([j for i in field_movies for j in i])
field_movies

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [41]:
# Show all fields in clean_movies_tara collection
field_tara = [list(doc.keys()) for doc in coll_tara.find({})]
field_tara = set([j for i in field_tara for j in i])
field_tara

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

- Dari kedua output di atas dapat dilihat bahwa **nama-nama fields pada movies dan clean_movies_tara sama persis**.

In [51]:
# The number of fields in movies collection
print('The number of fileds in movies collectin : ', len(field_movies))

The number of fileds in movies collectin :  22


In [52]:
# The number of fields in clean_movies_tara collection
print('The number of fileds in movies collectin : ', len(field_tara))

The number of fileds in movies collectin :  22


- Penjelasan di atas diperkuat dengan kedua output selanjutnya bahwa banyaknya fields pada movies dan clean_movies_tara sama, yaitu sebanyak **22 fields**.

## 6.3. All the values and order in both collections are the same

- Untuk melihat apakah values dan urutan pada movies dan clean_movies_tara sama, akan diambil sampel document yang memiliki ketentuan sama. Kemudian sampel documen tersebut akan diamati secara detail dan keseluruhan apakah isi dari kedua document tersebut sama persis atau tidak.

In [73]:
# Show document in movies collection where title = The Great Train Robbery, imdb.id = 439, imdb.rating = 7.4, imdb.votes = 9847
pprint.pprint(list(coll_movies.find({'title': 'The Great Train Robbery', 'imdb.id': 439, 'imdb.rating': 7.4, 'imdb.votes': 9847})))

[{'_id': ObjectId('573a1390f29313caabcd42e8'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['A.C. Abadie',
           "Gilbert M. 'Broncho Billy' Anderson",
           'George Barnes',
           'Justus D. Barnes'],
  'countries': ['USA'],
  'directors': ['Edwin S. Porter'],
  'fullplot': 'Among the earliest existing films in American cinema - notable '
              'as the first film that presented a narrative story to tell - it '
              'depicts a group of cowboy outlaws who hold up a train and rob '
              "the passengers. They are then pursued by a Sheriff's posse. "
              'Several scenes have color included - all hand tinted.',
  'genres': ['Short', 'Western'],
  'imdb': {'id': 439, 'rating': 7.4, 'votes': 9847},
  'languages': ['English'],
  'lastupdated': '2015-08-13 00:27:59.177000000',
  'plot': 'A group of bandits stage a brazen train hold-up, only to find a '
          'determined posse hot on their heels.',
  'poster': 'htt

In [72]:
# Show document in clean_movies_tara collection where title = The Great Train Robbery, imdb.id = 439, imdb.rating = 7.4, imdb.votes = 9847
pprint.pprint(list(coll_movies.find({'title': 'The Great Train Robbery', 'imdb.id': 439, 'imdb.rating': 7.4, 'imdb.votes': 9847})))

[{'_id': ObjectId('573a1390f29313caabcd42e8'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['A.C. Abadie',
           "Gilbert M. 'Broncho Billy' Anderson",
           'George Barnes',
           'Justus D. Barnes'],
  'countries': ['USA'],
  'directors': ['Edwin S. Porter'],
  'fullplot': 'Among the earliest existing films in American cinema - notable '
              'as the first film that presented a narrative story to tell - it '
              'depicts a group of cowboy outlaws who hold up a train and rob '
              "the passengers. They are then pursued by a Sheriff's posse. "
              'Several scenes have color included - all hand tinted.',
  'genres': ['Short', 'Western'],
  'imdb': {'id': 439, 'rating': 7.4, 'votes': 9847},
  'languages': ['English'],
  'lastupdated': '2015-08-13 00:27:59.177000000',
  'plot': 'A group of bandits stage a brazen train hold-up, only to find a '
          'determined posse hot on their heels.',
  'poster': 'htt

- Dari kedua output tersebut dapat dibandingkan bahwa **field** yang terdapat pada keduanya **sama**. Serta value pada masing-masing field nya juga sama.

## 7. Conclusion

- Terdapat 2 document yang ada pada movies namun tidak ada pada movies_initial.
- Terdapat 2 fields (tomatoes dan num_mflix_comments) yang ada pada movies namun tidak ada pada movies_initial.

<br>
Dengan kedua alasan itu sudah dapat dinyatakan bahwa kedua collection tersebut tidakan pernah sama persis. Namun disini sudah dibuat pipeline dengan ketentuan yang sama persis. Sehingga dapat disimpulkan bahwa field yang dimiliki sudah sama persis. Namun jumlah document tidak akan sama persis. Document yang ada pada clean_movies_tara pad ada pada movies. Namun 2 document pada movies tidak ada pada clean_movies_tara.